Added email analytics service

kevinansfield opened the pull request:
no issue

- added `EmailAnalyticsService`
  - `.fetchAll()` grabs and processes all available events
  - `.fetchLatest()` grabs and processes all events since the last seen event timestamp
  - `EventProcessor` passed event objects and updates `email_recipients` or `members` records depending on the event being analytics or list hygiene
    - always returns an `EventProcessingResult` instance so that progress can be tracked and merged across individual events, batches (pages of events), and total runs
    - adds email_id and member_id to the returned result where appropriate so that the stats aggregator can limit processing to data that has changed
    - sets `email_recipients.{delivered_at, opened_at, failed_at}` for analytics events
    - sets `members.subscribed = false` for permanent failure/unsubscribed/complained list hygiene events
  - `StatsAggregator` takes an `EventProcessingResult`-like object containing arrays of email ids and member ids on which to aggregate statistics.
  - jobs for `fetch-latest` and `fetch-all` ready for use with the JobsService
- added `initialiseRecurringJobs()` function to Ghost bootup procedure that schedules the email analytics "fetch latest" job to run every minute

TODO:
- [x] handle Mailgun not being set up
- [x] reset the Mailgun instance if related settings change

kevinansfield committed:
Added email analytics service

no issue

- added `EmailAnalyticsService`
  - `.fetchAll()` grabs and processes all available events
  - `.fetchLatest()` grabs and processes all events since the last seen event timestamp
  - `EventProcessor` passed event objects and updates `email_recipients` or `members` records depending on the event being analytics or list hygiene
    - always returns a `EventProcessingResult` instance so that progress can be tracked and merged across individual events, batches (pages of events), and total runs
    - adds email_id and member_id to the returned result where appropriate so that the stats aggregator can limit processing to data that has changed
    - sets `email_recipients.{delivered_at, opened_at, failed_at}` for analytics events
    - sets `members.subscribed = false` for permanent failure/unsubscribed/complained list hygiene events
  - `StatsAggregator` takes an `EventProcessingResult`-like object containing arrays of email ids and member ids on which to aggregate statistics.
  - jobs for `fetch-latest` and `fetch-all` ready for use with the JobsService
- added `initialiseRecurringJobs()` function to Ghost bootup procedure that schedules the email analytics "fetch latest" job to run every minute

naz commented on core/server/services/email-analytics/lib/event-processor.js:
```diff
@@ -0,0 +1,230 @@
+        this.providerIdEmailIdMap = {};
+    }
+
+    process(event) {
```
Think it's worth marking this method as `async` so it's clear for the caller that the result would be a Promise in most cases.

kevinansfield committed:
fixed empty settings cache when job is run

kevinansfield committed:
skip recurring jobs kick-off when testing

kevinansfield committed:
updated jobs to use parentPort

kevinansfield committed:
linting

kevinansfield committed:
match testing-* envs

kevinansfield committed:
let job manager handle errors

kevinansfield committed:
Revert "let job manager handle errors"

- job manager doesn't actually handle errors, jobs get stuck and the error output misses details and formatting that our logging util provides

kevinansfield committed:
fixed sub-query syntax

kevinansfield committed:
fixed missing await that was breaking member id fetching query

kevinansfield committed:
marked EventProcessor.process as async

kevinansfield committed:
skip fetch/aggregation when there are no emails in the system

kevinansfield committed:
made jobs less noisy and removed "maxEvents" timeboxing

- fetching a limited number of events poses problems when more than the limited number of events are generated within the Mailgun grace period, we'd end up re-fetching the same events over and over

kevinansfield committed:
added cancellation handling to analytics jobs

kevinansfield committed:
handle mailgun settings changing over time
